### DISCUSSION TITLE
General Comments - Block Protocol

### DISCUSSION THREAD

**Discussion by awsles** [blockprotocol/blockprotocol#118] [General]
Just took a look through the draft block protocol spec. Nice first cut. Is the block protocol about data _interchange_ or about data _presentation_ (in a UI)? A few comments:

•	Some simple (complete – not excerpts) examples are needed within the specification. A pointer to a hash.ai github repo where you have to do a treasure hunt is not helpful. 😊 Even in those cases, at least provide a link to a specific file.
•	Also provide excerpt examples. You do a few later in the doc, but do these early.  For example, what is the basic structure? 
•	I’m a little fuzzy about why there is embedded code?  If there is, then are you defining an object (vs. a data interchange)?
•	Hard-coded filenames are a non-starter (block-schema.json and block-metadata.json). For one, what if I want multiple definitions in the same directory? There would be collisions. These should simply be input parameters into whatever is consuming them. 
•	Ontology is conspicuously missing…  data sets tie to ontology. It should at least be discussed. Ontology helps with the meaning of the data.
•	Define “Entity”… Add a section on “Entities” and discuss what they are. The spec jumps right into functions without any context.
•	Block permissions – I suggest staying away from this here entirely.  For one, the blocks move around and therefore can be edited en route (okay, you could digitally sign permissions but what is the point if the data is embedded? Encrypted you say?  Okay this is boiling the ocean now.  😊)  I suggest avoiding security for now and focusing on interchange. 
•	User Actions – Why? This is outside the requirements for interchange. I’d opt for keeping things simple. Simple is better.
•	Comments Facility – This is up to the data owner. Could be embedded (if they want it to be part of the data set) or could be a wrapper.
•	Data Types – Where are these defined? 
•	What is the embedded code for?  Data should be code agnostic.
•	500 Internal Server Error on the link to"rendering context": https://blockprotocol.org/spec/implementation-and-rendering-context 😱

Having read the spec, I still don’t know what a block looks like. A basic example is needed for each block type. There are some references to HTML parts… but I get the sense that the block intends to have self-described data which can readily be rendered.

**Comment by CiaranMn** [comment]
Hi @leswaters - thank you for the detailed feedback, much appreciated!

Taking your points in turn:

- Some simple (complete – not excerpts) examples are needed within the specification. A pointer to a hash.ai github repo where you have to do a treasure hunt is not helpful. 😊 Even in those cases, at least provide a link to a specific file.
- Also provide excerpt examples. You do a few later in the doc, but do these early. For example, what is the basic structure?
- Having read the spec, I still don’t know what a block looks like. A basic example is needed for each block type. There are some references to HTML parts… but I get the sense that the block intends to have self-described data which can readily be rendered.

Agree - I've added one link and will add more. We will add more excerpts also.

- I’m a little fuzzy about why there is embedded code? If there is, then are you defining an object (vs. a data interchange)?
- What is the embedded code for? Data should be code agnostic.

I may have misunderstood these questions, but: we are defining a data interchange between an embedding application and a block, in a web context, and for that reason there has to be some discussion of the implementation detail of how data and requests are passed back and forth between the application and block (i.e. in a browser). While the data itself is code agnostic, the method of moving it around in this context is less so.

- Hard-coded filenames are a non-starter (block-schema.json and block-metadata.json). For one, what if I want multiple definitions in the same directory? There would be collisions. These should simply be input parameters into whatever is consuming them.

We can have a think about this - ultimately there needs to be a known entry point into the package of files which can then point to other files. The entry point can either be known by file naming convention, or stored some other way. File naming convention for 'here is the file describing my library' is pretty standard in many languages/contexts. The block will be a number of source files, so we'd expect people to keep them in separate directories.

- Ontology is conspicuously missing… data sets tie to ontology. It should at least be discussed. Ontology helps with the meaning of the data.

We have [added a FAQ](https://blockprotocol.org/docs/faq) which touches on this - does it help?

- Define “Entity”… Add a section on “Entities” and discuss what they are. The spec jumps right into functions without any context.

We have something (albeit brief) [here](https://blockprotocol.org/spec#terminology).

- Block permissions – I suggest staying away from this here entirely. For one, the blocks move around and therefore can be edited en route (okay, you could digitally sign permissions but what is the point if the data is embedded? Encrypted you say? Okay this is boiling the ocean now. 😊) I suggest avoiding security for now and focusing on interchange.
- User Actions – Why? This is outside the requirements for interchange. I’d opt for keeping things simple. Simple is better.
- Comments Facility – This is up to the data owner. Could be embedded (if they want it to be part of the data set) or could be a wrapper.

There's a general principle we're thinking of in touching on these, which is that as well as standardising the interchange between applications and blocks, we're also focusing (at least at first) on how these operate in a web editing context - e.g. content management systems - and want to at least address a few obvious things arising from this. I agree we do not want to be in the business of prescribing how security is managed - just gently suggest that applications might like to think carefully about it. 🙂 

I agree that limiting the scope makes the endeavour more manageable, and simpler is better where possible.

- Data Types – Where are these defined?

Briefly [here](https://blockprotocol.org/spec#terminology) - unless you mean something other than block types?

- 500 Internal Server Error on the link to"rendering context": https://blockprotocol.org/spec/implementation-and-rendering-context 😱

Good spot, fixed!

Thanks again for your time and thought in contributing - I welcome any further thoughts you may have.
